### DISCUSSION TITLE
BLE Gamepad esp32

### DISCUSSION THREAD

**Discussion by salarizadi** [Moddable-OpenSource/moddable#1277] [Q&A]
I am making a gamepad. I want to make a gamepad for my mobile phone with esp32 that can be connected to my mobile phone with bluetooth, but I did not fully understand how I can send the necessary codes or letters from esp32 to my mobile phone with bluetooth and perform the operation. If you can write a sample or send a project sample link, I would be grateful 🙏

**Comment by mkellner** [comment]
That sounds like a great project!

There is a blog post with sample projects that can be found here:

https://www.moddable.com/blog/bleremotes/

**Comment by salarizadi** [comment]
I understood and did all the Bluetooth steps. When I want to find my mobile in the Bluetooth list with esp32, unfortunately, it cannot find my mobile, only my laptop and my watch. What is the reason?

**Reply by salarizadi** [reply]
@mkellner @phoddie

**Reply by mkellner** [reply]
@salarizadi, is there any bluetooth device that will control your mobile?

**Reply by salarizadi** [reply]
I didn't understand what you meant, but I tested esp32-wroom with two different versions and uploaded the sample scanner codes. It did not find my phone. My phone model is Honor 20, do I need to make special settings?

I also tested with the Arduino code, but it did not find my mobile either

**Reply by phoddie** [reply]
@salarizadi – It may be that your phone does have the necessary BLE service support for this use. But, it is difficult to say because I'm not sure what you have tried. Please let us know (1) the application you have installed on the ESP32 and (2) how you connected to that from your computer, and (3) how you tried to connect to it from your phone. Thank you.

**Reply by salarizadi** [reply]
I connected the esp32 to the laptop with a usb cable to upload. I use the CH340 programmer module to upload.

1) I have uploaded sample code of BLEClient bluetooth scanner in esp32
2) I turned on the bluetooth of my mobile phone
3) I waited for it to show me all nearby bluetooth devices
4) I saw that it does not show Bluetooth
5) I turned on the bluetooth of the watch and the laptop
6) And it only showed the bluetooth of my laptop and watch

Bluetooth doesn't show my phone just among all bluetooth

**Reply by salarizadi** [reply]
@phoddie 

From the sample code:

https://github.com/Moddable-OpenSource/moddable/tree/public/examples%2Fnetwork%2Fble%2Fscanner

Use it and see if you can find your mobile, maybe the problem is with my phone and not esp32

**Reply by phoddie** [reply]
The scanner app just looks for devices that are transmitting BLE advertising packets. We've seen that work with both Android and iPhone. From what you report, your phone may not be discoverable. That's consistent with your observation that it also doesn't work with your Arduino test.

That said, this is something of a side-trip. The blog post @mkellner linked to works in the opposite direction. That is to say, the ESP32 is advertising the HID keyboard and then computer connects to that.

**Reply by salarizadi** [reply]
@phoddie Your words are correct and logical, but why can the laptop find my phone, why can't the esp32. I also tested the BLE server and my phone found esp32. And what if I want to build a whole system that shows all bluetooth? I want to scan all bluetooth around me with esp32, is this possible?

**Reply by phoddie** [reply]
> ....why can the laptop find my phone, why can't the esp32

I don't know. There are many variables there.

> And what if I want to build a whole system that shows all bluetooth? I want to scan all bluetooth around me with esp32, is this possible?

That is what the scanner does.

I understand it is frustrating that the ESP32 does not see your phone. There is no information here that indicates that is a problem in the Moddable SDK.

**Reply by salarizadi** [reply]
@phoddie Thanks for being so patient and quick to answer me, I'm still trying to understand how to scan more accurately and will let you know. Thank you very much.

**Reply by salarizadi** [reply]
@phoddie 

I was finally able to connect my mobile to esp32 and I was even able to control my mobile but I have a new problem, I have been searching the whole internet for several hours, unfortunately I did not understand how to send the DPAD_UP, DPAD_DOWN, DPAD_RIGHT and DPAD_LEFT buttons correctly and that I tested the gamepad I'm making on the Asphalt 9 game, unfortunately it can't recognize that it's a gamepad and it just says that the keyboard is turned on. What should I do so that it can understand that a gamepad is connected and not a keyboard?

I even set the appearance option to 964. In the Bluetooth list, it showed the gamepad icon, but in the game it still says that you are connected to a keyboard.

**Reply by salarizadi** [reply]
@phoddie I added some new HID codes and was able to send them, but unfortunately HID does not recognize the DPAD code :

```js
const HID_ARROWS = {
    ArrowUp   : 0x52, // Success
    ArrowDown : 0x51, // Success
    ArrowRight: 0x4f, // Success
    ArrowLeft : 0x50, // Success

    DPAD_UP   : 0x13, // Failed
    DPAD_DOWN : 0x14, // Failed
    DPAD_LEFT : 0x15, // Failed
    DPAD_RIGHT: 0x16 // Failed
};
```

**Reply by phoddie** [reply]
> I was finally able to connect my mobile to esp32 

Excellent. Would you please share your solution? It may help others.

> unfortunately I did not understand how to send the DPAD_UP, DPAD_DOWN, DPAD_RIGHT and DPAD_LEFT buttons correctly.... What should I do so that it can understand that a gamepad is connected and not a keyboard?

I'm not sure. The example reports that it is a keyboard. That starts in [hid.js](https://github.com/Moddable-OpenSource/moddable/blob/public/contributed/ble/hid-peripheral/host/bleservices/hid.json). To adapt the example to work with the game, you need to know the BLE profile for the gamepad. I don't know that. Maybe @andycarle can help,

**Reply by salarizadi** [reply]
@phoddie

> Excellent. Would you please share your solution? It may help others.

When the project is finished, I'll upload all the code with descriptions to my GitHub page 😉

**Reply by andycarle** [reply]
@salarizadi As @phoddie notes, you are heading into an area that requires a real understanding of the HID specification. To accomplish your goals, you must update the HID report map and descriptors in [hid.js](https://github.com/Moddable-OpenSource/moddable/blob/public/contributed/ble/hid-peripheral/host/bleservices/hid.json) to contain the device types and buttons/keys/joysticks/etc. that you wish to report from your ESP32 to your phone. This is not a small project, but it is very doable.

The starting point for this work is the (rather archaic) HID Descriptor Tool. That is available from the USB Implementors Forum [here](https://usb.org/document-library/hid-descriptor-tool). You can use that tool to update the HID report map and descriptor I included in the HID Host that powers the BLE Remotes examples you have been using. That descriptor is documented in the [HID Peripheral Host documentation's HID Report Descriptor section](https://github.com/Moddable-OpenSource/moddable/blob/public/contributed/ble/hid-peripheral/README.md#hid-report-descriptor).

I can help you further down this path (I have done much of this work before for one-off projects), but for you to build a custom solution, you'll really need to understand this stuff yourself. I recommend spending some time with the HID specification documents, forum posts about HID descriptors, and HID sample code from other (not Moddable SDK) projects. Nothing that you need to change is specific to the Moddable SDK, JavaScript, or even BLE--it's all core USB HID stuff from the mid-'90s.

**Reply by salarizadi** [reply]
@andycarle @phoddie 

happy New Year 🎊 good news!
Finally, I was able to make the phone and other devices recognize the esp32 as a gamepad :)

But unfortunately...no matter what I do, it doesn't send the codes through Bluetooth. I'm still working on this problem, I didn't think it would be so much trouble to make a gamepad 🤦‍♂️

**Reply by andycarle** [reply]
@salarizadi What do your HID input reports look like? It may be that your input report doesn't match the length specified by your HID descriptor. Or that you have a mismatch between the report ID referenced in the descriptor and that report ID used for the input report.

**Reply by salarizadi** [reply]
@andycarle Unfortunately I couldn't fix it, I did everything I could, tested more than 20 HID report maps, cleaned the code several times, read the old 90's docs and the new docs, but unfortunately I couldn't fix it 🤦‍♂️
